File: ios/Stickers/OpenMojiCatalog.swift
import Foundation

/// A single entry from OpenMoji's `openmoji.json` index that we care about.
/// Mirrors Android's `OpenMojiCatalog.Entry` in spirit (same field set).
struct OpenMojiEntry {
  let emoji: String
  let hexcode: String
  let group: String
  let annotation: String
  let tags: String
}

/// Fetches, disk-caches, and serves the OpenMoji sticker index
/// (`https://cdn.jsdelivr.net/gh/hfg-gmuend/openmoji@master/data/openmoji.json`).
///
/// No third-party networking/JSON dependency is used — this codebase is deliberately
/// dependency-light (see `stickerAssets: [[String: Any]]` in `PhotoVideoEditorViewController`
/// for the existing dictionary-based JSON precedent). Plain `URLSession` + `JSONSerialization`.
final class OpenMojiCatalog {
  static let shared = OpenMojiCatalog()

  /// Fixed, sensible display order for OpenMoji's `group` values.
  static let orderedGroups: [String] = [
    "smileys-emotion",
    "people-body",
    "animals-nature",
    "food-drink",
    "travel-places",
    "activities",
    "objects",
    "symbols",
    "flags",
  ]

  private static let indexURL = URL(string: "https://cdn.jsdelivr.net/gh/hfg-gmuend/openmoji@master/data/openmoji.json")!
  private static let cacheFileName = "openmoji_index.json"
  private static let cacheTimestampKey = "OpenMojiCatalog.cacheTimestamp"
  private static let cacheMaxAgeSeconds: TimeInterval = 7 * 24 * 60 * 60 // 7 days

  private let queue = DispatchQueue(label: "com.photovideoeditor.openmojicatalog", qos: .userInitiated)
  private var entries: [OpenMojiEntry]?
  private var isLoading = false
  private var pendingCompletions: [([OpenMojiEntry]?, Error?) -> Void] = []

  private init() {}

  private var cacheFileURL: URL? {
    FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?
      .appendingPathComponent(Self.cacheFileName)
  }

  enum CatalogError: Error {
    case network(Error)
    case badResponse
    case emptyCache
  }

  /// Loads the catalog (cache-first, network fallback with cache-refresh) and calls `completion`
  /// on the main thread. Safe to call multiple times concurrently; duplicate in-flight calls are coalesced.
  func load(completion: @escaping ([OpenMojiEntry]?, Error?) -> Void) {
    if let entries {
      DispatchQueue.main.async { completion(entries, nil) }
      return
    }
    queue.async { [weak self] in
      guard let self else { return }
      self.pendingCompletions.append(completion)
      if self.isLoading { return }
      self.isLoading = true
      self.loadLocked()
    }
  }

  /// Must be called on `queue`. Decides whether to use the on-disk cache or fetch fresh, then
  /// fans results out to every completion queued while the load was in flight.
  private func loadLocked() {
    let cacheAge = self.cacheAgeSeconds()
    if let cacheAge, cacheAge < Self.cacheMaxAgeSeconds, let cached = self.readCache() {
      self.finishLoad(entries: cached, error: nil)
      return
    }
    self.fetchFromNetwork { [weak self] data, error in
      guard let self else { return }
      self.queue.async {
        if let data {
          self.writeCache(data)
          let parsed = self.parse(data)
          self.finishLoad(entries: parsed, error: parsed.isEmpty ? CatalogError.badResponse : nil)
        } else if let cached = self.readCache() {
          // Network failed but we have a cache of any age — fall back to it.
          self.finishLoad(entries: cached, error: nil)
        } else {
          self.finishLoad(entries: nil, error: error ?? CatalogError.emptyCache)
        }
      }
    }
  }

  /// Must be called on `queue`.
  private func finishLoad(entries: [OpenMojiEntry]?, error: Error?) {
    if let entries { self.entries = entries }
    let completions = self.pendingCompletions
    self.pendingCompletions = []
    self.isLoading = false
    DispatchQueue.main.async {
      completions.forEach { $0(entries, entries == nil ? error : nil) }
    }
  }

  private func fetchFromNetwork(completion: @escaping (Data?, Error?) -> Void) {
    let task = URLSession.shared.dataTask(with: Self.indexURL) { data, response, error in
      if let error {
        completion(nil, CatalogError.network(error))
        return
      }
      guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode), let data else {
        completion(nil, CatalogError.badResponse)
        return
      }
      completion(data, nil)
    }
    task.resume()
  }

  private func cacheAgeSeconds() -> TimeInterval? {
    let timestamp = UserDefaults.standard.double(forKey: Self.cacheTimestampKey)
    guard timestamp > 0 else { return nil }
    return Date().timeIntervalSince1970 - timestamp
  }

  private func readCache() -> [OpenMojiEntry]? {
    guard let url = cacheFileURL, let data = try? Data(contentsOf: url) else { return nil }
    let parsed = parse(data)
    return parsed.isEmpty ? nil : parsed
  }

  private func writeCache(_ data: Data) {
    guard let url = cacheFileURL else { return }
    do {
      try data.write(to: url, options: .atomic)
      UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: Self.cacheTimestampKey)
    } catch {
      // Best-effort cache; ignore write failures.
    }
  }

  private func parse(_ data: Data) -> [OpenMojiEntry] {
    guard let raw = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else { return [] }
    return raw.compactMap { dict in
      guard
        let hexcode = dict["hexcode"] as? String, !hexcode.isEmpty,
        let group = dict["group"] as? String
      else { return nil }
      let emoji = dict["emoji"] as? String ?? ""
      let annotation = dict["annotation"] as? String ?? ""
      let tags = dict["tags"] as? String ?? ""
      return OpenMojiEntry(emoji: emoji, hexcode: hexcode, group: group, annotation: annotation, tags: tags)
    }
  }

  // MARK: - Query surface

  func groups() -> [String] {
    guard let entries else { return [] }
    let present = Set(entries.map { $0.group })
    return Self.orderedGroups.filter { present.contains($0) }
  }

  func byGroup(_ group: String) -> [OpenMojiEntry] {
    (entries ?? []).filter { $0.group == group }
  }

  func search(_ query: String) -> [OpenMojiEntry] {
    let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard !needle.isEmpty else { return entries ?? [] }
    return (entries ?? []).filter {
      $0.annotation.lowercased().contains(needle) || $0.tags.lowercased().contains(needle)
    }
  }

  func thumbnailURL(_ hexcode: String) -> URL {
    URL(string: "https://cdn.jsdelivr.net/gh/hfg-gmuend/openmoji@master/color/72x72/\(hexcode).png")!
  }

  func fullURL(_ hexcode: String) -> URL {
    URL(string: "https://cdn.jsdelivr.net/gh/hfg-gmuend/openmoji@master/color/618x618/\(hexcode).png")!
  }
}

File: android/src/main/java/com/photovideoeditor/stickers/OpenMojiCatalog.kt
package com.photovideoeditor.stickers

import android.content.Context
import android.os.Handler
import android.os.Looper
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executors

/** One entry from OpenMoji's `data/openmoji.json` index. */
data class OpenMojiEntry(
  val emoji: String,
  val hexcode: String,
  val group: String,
  val annotation: String,
  val tags: String
)

/**
 * Fetches and disk-caches the OpenMoji catalog index (CC BY-SA 4.0, https://openmoji.org),
 * exposing search/filter helpers plus the CDN URL builders for thumbnail/full-res assets.
 *
 * Networking uses plain [HttpURLConnection] on a background executor (this codebase has no
 * networking dependency and shouldn't gain one for a simple GET) and parses with
 * [org.json.JSONArray]/[org.json.JSONObject], matching the JSON style already used elsewhere
 * (e.g. `PhotoVideoEditorActivity`'s `stickerAssets` parsing).
 */
object OpenMojiCatalog {
  private const val INDEX_URL =
    "https://cdn.jsdelivr.net/gh/hfg-gmuend/openmoji@master/data/openmoji.json"
  private const val CACHE_FILE_NAME = "openmoji_index.json"
  private const val CACHE_TIMESTAMP_PREFS = "openmoji_cache"
  private const val CACHE_TIMESTAMP_KEY = "fetched_at"
  private val CACHE_TTL_MS = 7L * 24L * 60L * 60L * 1000L

  /** Fixed, sensible display order for the "All" chip row; unknown groups are appended after. */
  private val GROUP_DISPLAY_ORDER = listOf(
    "smileys-emotion",
    "people-body",
    "animals-nature",
    "food-drink",
    "travel-places",
    "activities",
    "objects",
    "symbols",
    "flags"
  )

  private val executor = Executors.newSingleThreadExecutor { task -> Thread(task, "openmoji-catalog") }
  private val mainHandler = Handler(Looper.getMainLooper())

  @Volatile private var cachedEntries: List<OpenMojiEntry>? = null

  /**
   * Loads the catalog (memory -> disk cache -> network, falling back to any cached copy on
   * network failure), delivering the result on the main thread.
   */
  fun load(context: Context, onResult: (entries: List<OpenMojiEntry>?, error: Exception?) -> Unit) {
    val inMemory = cachedEntries
    if (inMemory != null) {
      mainHandler.post { onResult(inMemory, null) }
      return
    }
    val appContext = context.applicationContext
    executor.execute {
      try {
        val entries = loadBlocking(appContext)
        cachedEntries = entries
        mainHandler.post { onResult(entries, null) }
      } catch (e: Exception) {
        mainHandler.post { onResult(null, e) }
      }
    }
  }

  private fun loadBlocking(context: Context): List<OpenMojiEntry> {
    val cacheFile = File(context.cacheDir, CACHE_FILE_NAME)
    val cacheAgeMs = cacheAgeMs(context)
    val cacheFresh = cacheFile.exists() && cacheAgeMs != null && cacheAgeMs < CACHE_TTL_MS

    if (cacheFresh) {
      return parse(cacheFile.readText())
    }

    return try {
      val json = fetch(INDEX_URL)
      cacheFile.writeText(json)
      markCacheTimestamp(context)
      parse(json)
    } catch (e: Exception) {
      // Network failed: fall back to any cached copy, regardless of age, per spec.
      if (cacheFile.exists()) parse(cacheFile.readText()) else throw e
    }
  }

  private fun fetch(urlString: String): String {
    val connection = URL(urlString).openConnection() as HttpURLConnection
    connection.instanceFollowRedirects = true
    connection.connectTimeout = 15_000
    connection.readTimeout = 30_000
    connection.requestMethod = "GET"
    try {
      val code = connection.responseCode
      if (code !in 200..299) throw java.io.IOException("HTTP $code fetching $urlString")
      return connection.inputStream.bufferedReader().use { it.readText() }
    } finally {
      connection.disconnect()
    }
  }

  private fun parse(json: String): List<OpenMojiEntry> {
    val array = JSONArray(json)
    val result = ArrayList<OpenMojiEntry>(array.length())
    for (i in 0 until array.length()) {
      val obj: JSONObject = array.optJSONObject(i) ?: continue
      val hexcode = obj.optString("hexcode").takeIf { it.isNotBlank() } ?: continue
      result.add(
        OpenMojiEntry(
          emoji = obj.optString("emoji"),
          hexcode = hexcode,
          group = obj.optString("group"),
          annotation = obj.optString("annotation"),
          tags = obj.optString("tags")
        )
      )
    }
    return result
  }

  private fun cacheAgeMs(context: Context): Long? {
    val prefs = context.getSharedPreferences(CACHE_TIMESTAMP_PREFS, Context.MODE_PRIVATE)
    val fetchedAt = prefs.getLong(CACHE_TIMESTAMP_KEY, -1L)
    if (fetchedAt < 0) return null
    return System.currentTimeMillis() - fetchedAt
  }

  private fun markCacheTimestamp(context: Context) {
    context.getSharedPreferences(CACHE_TIMESTAMP_PREFS, Context.MODE_PRIVATE)
      .edit()
      .putLong(CACHE_TIMESTAMP_KEY, System.currentTimeMillis())
      .apply()
  }

  fun groups(): List<String> {
    val entries = cachedEntries ?: return emptyList()
    val present = entries.map { it.group }.distinct()
    val ordered = GROUP_DISPLAY_ORDER.filter { it in present }
    val rest = present.filter { it !in GROUP_DISPLAY_ORDER }.sorted()
    return ordered + rest
  }

  fun byGroup(group: String): List<OpenMojiEntry> =
    cachedEntries?.filter { it.group == group } ?: emptyList()

  fun search(query: String): List<OpenMojiEntry> {
    val entries = cachedEntries ?: return emptyList()
    val needle = query.trim().lowercase()
    if (needle.isEmpty()) return entries
    return entries.filter {
      it.annotation.lowercase().contains(needle) || it.tags.lowercase().contains(needle)
    }
  }

  fun thumbnailUrl(hexcode: String): String =
    "https://cdn.jsdelivr.net/gh/hfg-gmuend/openmoji@master/color/72x72/$hexcode.png"

  fun fullUrl(hexcode: String): String =
    "https://cdn.jsdelivr.net/gh/hfg-gmuend/openmoji@master/color/618x618/$hexcode.png"
}
